### PR TITLE
Inject the conduit proxy into controller pod during conduit install

### LINKED ISSUE #311: Inject the Conduit proxy into all Conduit control plane pods
The control plane itself should have all the observability, reliability, and security benefits that any Conduit-managed service would have. For example, we should be able to see request success rates and latencies for the private API and the public API and for the Web UI in the Conduit web UI and in `conduit stats`. Further, the way the control plane is secured (via TLS, in particular) should be done in the same way it is done for the data plane. To do this without duplicating work between the control plane code (in Go) and the proxy code (in Rust), the proxy should be injected into all the control plane pods and as much as possible everything should work as though the control plane is part of the data plane.

There may be bootstrapping-specific stuff we need to address since the proxy will both be proxying the control plane and using the control plane's private API as a client. See #302.

/cc @olix0r @adleong

### PR DESCRIPTION
**Pull request by deebo91**
In order to take advantage of the benefits the conduit proxy gives to deployments, this PR injects the conduit proxy into the control plane pod. This helps us lay the groundwork for future work such as TLS, control plane observability etc.

Fixes #311

Signed-off-by: Dennis Adjei-Baah dennis@buoyant.io

### CODE REVIEW

File: cli/cmd/testdata/deployment_test_file.golden.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
+apiVersion: apps/v1beta1
```
Comment (briansmith) [inline cli/cmd/testdata/deployment_test_file.golden.yml]: I think this file should start with the test name, "TestInjectYAML", according to, e.g., https://medium.com/soon-london/testing-with-golden-files-in-go-7fccc71c43d3.

Also, it is likely that we'll have multiple tests for this, so maybe include "emojivoto" in the name.

I don't think "test_file" is a useful part of the filename. I suggest dropping it.

Wrapping up, I suggest this file be named "TestInjectYAML_emojivoto_deployment.golden.yml"

File: cli/cmd/testdata/deployment_test_file.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+---
```
Comment (briansmith) [inline cli/cmd/testdata/deployment_test_file.yml]: Using similar logic, this should be named "TestInjectYAML_emojivoto_deployment.input.yml". In particular it is useful to have "input" in the file name.

File: cli/cmd/install.go [hunk 3]
```diff
@@ -442,7 +445,11 @@ var installCmd = &cobra.Command{
 			ControllerComponentLabel: k8s.ControllerComponentLabel,
 			CreatedByAnnotation:      k8s.CreatedByAnnotation,
 		}
-		return render(config, os.Stdout)
+		err := render(config, buf)
+		if err != nil {
+			return err
+		}
+		return InjectYAML(buf, os.Stdout)
```
Comment (briansmith) [inline cli/cmd/install.go]: This is OK as long as we fix #303 in the same release that this lands in (i.e. either we need to fix #303 in for conduit 0.3, or we need to defer this for 0.4 and fix #303 in 0.4). In particular, we really don't want `conduit install` to output a bad or imcomplete configuration if InjectYAML fails partway through. However, we can resolve #303 in a separate PR.

Comment (deebo91) [reply cli/cmd/install.go]: I can open a seperate PR for this right after I make these changes.

**Comment by briansmith** [comment]
> We need to add TestConduitInstall unit test that includes a golden file of what we expect the default output of conduit install to be. Then the reviewer (myself) can verify that the output is the same as conduit install | conduit inject and we'll be able to see in future reviews how each change to conduit install changes that output.

Ideally we should add this `TestConduitInstall` test as a separate PR, and rebase this PR on top of that one. That way this PR will include a diff of exactly what changes in the output of `conduit install` after these code changes.

**Comment by deebo91** [comment]
@briansmith we have `TestRender` under `install_test.go` which has the bulk of the install code that generate which has a stripped down version of a golden test file. Does that suffice as a test for the install command?

**Comment by briansmith** [comment]
> @briansmith we have TestRender under install_test.go which has the bulk of the install code that generate which has a stripped down version of a golden test file. Does that suffice as a test for the install command?

To be a sufficient test, this PR should have had changes to testdata/install_output.golden. Since this PR didn't change install_output.golden, I think we either need a new test or we need to change TestRender and/or the implementation of of your code changes here. I don't care which strategy is used.

File: cli/cmd/inject_test.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		file, err := os.Open("testdata/inject_emojivoto_deployment.input.yml")
+		if err != nil {
+			t.Errorf("error opening test file: %v\n", err)
+		}
+
+		goldenFileBytes, err := ioutil.ReadFile("testdata/inject_emojivoto_deployment.golden.yml")
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		expectedContent := string(goldenFileBytes)
```
Comment (briansmith) [inline cli/cmd/inject_test.go]: NIT 1: It seems better to me to read the golden file later, after we've done the "Unexpected error" test and right above the check that the expected output is the same as the actual output.

NIT 2: This variable is named `expectedContent` and but the other variable is named `actualOutput`. I suggest using parallel naming: `actualOutput` and `expectedOutput`.

File: cli/cmd/inject_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		output := new(bytes.Buffer)
+
+		err = InjectYAML(read, output)
+		if err != nil {
+			t.Errorf("Unexpected error injecting YAML: %v\n", err)
+		}
+
+		actualOutput := output.String()
+
+		if actualOutput != expectedContent {
```
Comment (briansmith) [inline cli/cmd/inject_test.go]: The logic for reading a golden file (lines 20-23) and this code for comparing the expected results equal the actual results and outputing the diff and failing if they don't, is all exactly the same between the various tests. We should combine this shared logic into a helper function that all such tests use, rather than copy-pasting it into each test.

File: cli/cmd/testdata/install_output.golden [hunk 6]
```diff
@@ -117,101 +117,164 @@ spec:
     targetPort: 8086
 
 ---
-kind: Deployment
 apiVersion: extensions/v1beta1
+kind: Deployment
```
Comment (briansmith) [inline cli/cmd/testdata/install_output.golden]: I filed #372 regarding the fact that `conduit inject` perturbs the input unnecessarily like this. No action is needed to undo this in this PR.

File: cli/cmd/testdata/install_output.golden [hunk 7]
```diff
@@ -117,101 +117,164 @@ spec:
...
+        - containerPort: 8087
+          name: grpc
+        - containerPort: 9997
+          name: admin-http
+        resources: {}
+      - env:
+        - name: CONDUIT_PROXY_LOG
+          value: warn,conduit_proxy=info
+        - name: CONDUIT_PROXY_CONTROL_URL
+          value: tcp://proxy-api.conduit.svc.cluster.local:8086
```
Comment (adleong) [inline cli/cmd/testdata/install_output.golden]: As we discussed, I think this should be localhost

Comment (briansmith) [reply cli/cmd/testdata/install_output.golden]: > As we discussed, I think this should be localhost

As we discussed in the meeting, it's better for that that to happen in a separate PR that gets merged immediately after this one, so that we can compare the behavior of the localhost vs. proxy-api.conduit.svc.cluster.local configuration and so that the change to the golden file for the `localhost` is isolated.

File: cli/cmd/test_helper.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+func diffCompare(t *testing.T, actual string, expected string) {
+	if actual != expected {
+		dmp := diffmatchpatch.New()
+		diffs := dmp.DiffMain(actual, expected, true)
+		patches := dmp.PatchMake(expected, diffs)
+		patchText := dmp.PatchToText(patches)
+		t.Fatalf("Unexpected output:\n%+v", patchText)
+
```
Comment (briansmith) [inline cli/cmd/test_helper.go]: NIT: The extra blank line here should be removed.

Comment (deebo91) [reply cli/cmd/test_helper.go]: Oops sorry about that.

**Comment by deebo91** [comment]
@briansmith I am curious to know, what is the priority order for the related PRs?

**Comment by briansmith** [comment]
> @briansmith I am curious to know, what is the priority order for the related PRs?

I would fix #303 first since that's a bug independent of this issue.

Then I would verify if `conduit tap` works and more generally whether the other integration tests work, and fix things if necessary.

Then I would do the `localhost` issue. (This might be part of getting the integration tests to pass.)

**Comment by deebo91** [comment]
Looks like I was mistaken about the tap command. It works with the DNS name. I am still going to be pushing up the localhost address PR though.

```bash
➜  conduit git:(deebo91/inject-proxy-into-control-plane) go run cli/main.go tap deployment emojivoto/vote-bot
req id=0:100692 src=10.1.0.201:48318 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/list
rsp id=0:100692 src=10.1.0.201:48318 dst=10.1.0.199:80 :status=200 latency=31338µs
end id=0:100692 src=10.1.0.201:48318 dst=10.1.0.199:80 duration=560µs response-length=4558B
req id=0:100693 src=10.1.0.201:48318 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/vote
rsp id=0:100693 src=10.1.0.201:48318 dst=10.1.0.199:80 :status=200 latency=19878µs
end id=0:100693 src=10.1.0.201:48318 dst=10.1.0.199:80 duration=13µs response-length=0B
req id=0:100694 src=10.1.0.201:48318 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/list
rsp id=0:100694 src=10.1.0.201:48318 dst=10.1.0.199:80 :status=200 latency=18540µs
end id=0:100694 src=10.1.0.201:48318 dst=10.1.0.199:80 duration=1079µs response-length=4558B
req id=0:100695 src=10.1.0.201:48318 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/vote
rsp id=0:100695 src=10.1.0.201:48318 dst=10.1.0.199:80 :status=500 latency=12915µs
end id=0:100695 src=10.1.0.201:48318 dst=10.1.0.199:80 duration=404µs response-length=51B
req id=0:100696 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/list
rsp id=0:100696 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=200 latency=25571µs
end id=0:100696 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=909µs response-length=4558B
req id=0:100697 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/vote
rsp id=0:100697 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=200 latency=11906µs
end id=0:100697 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=31µs response-length=0B
req id=0:100698 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/list
rsp id=0:100698 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=200 latency=10505µs
end id=0:100698 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=167µs response-length=4558B
req id=0:100699 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/vote
rsp id=0:100699 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=200 latency=15608µs
end id=0:100699 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=14µs response-length=0B
req id=0:100700 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/list
rsp id=0:100700 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=200 latency=14331µs
end id=0:100700 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=984µs response-length=4558B
req id=0:100701 src=10.1.0.201:48436 dst=10.1.0.199:80 :method=GET :authority=web-svc.emojivoto:80 :path=/api/vote
rsp id=0:100701 src=10.1.0.201:48436 dst=10.1.0.199:80 :status=500 latency=14692µs
end id=0:100701 src=10.1.0.201:48436 dst=10.1.0.199:80 duration=512µs response-length=51B
```

File: cli/cmd/install.go [hunk 9]
```diff
@@ -451,7 +453,12 @@ func render(config installConfig, w io.Writer) error {
 	if err != nil {
 		return err
 	}
-	return template.Execute(w, config)
+	buf := new(bytes.Buffer)
```
Comment (siggy) [inline cli/cmd/install.go]: nit: consider stack allocation `buf := &bytes.Buffer{}`

File: cli/cmd/install_test.go [hunk 10]
```diff
@@ -31,7 +31,6 @@ func TestRender(t *testing.T) {
 			ControllerComponentLabel: "ControllerComponentLabel",
 			CreatedByAnnotation:      "CreatedByAnnotation",
 		}
-
```
Comment (briansmith) [inline cli/cmd/install_test.go]: NIT: Remove this removal of a blank line that was there before.

**Comment by briansmith** [comment]
> it looks minikube doesn't play nice with the DNS name

What are the symptoms?

> I think the localhost change needs to be merged directly after

I agree that we should do the localhost change regardless. However, it would be useful to know what is breaking minikube specifically since that seems surprising. From what I understand nothing should break, and the localhost change is not to fix anything but rather to avoid some overhead in metrics and networking, so if something is breaking now I'd like to know better what it is before we wallpaper over it with the localhost change.

**Comment by deebo91** [comment]
> What are the symptoms?

Symptoms seem to be exactly like #302 where `nghttp` the public or proxy api from a container within hangs indefinitely
```bash
root@controller-799c7d4ffc-zzxmr:/# nghttp -v http://api.conduit.svc.cluster.local:8086/api/v1/Report
[  0.001] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.001] send HEADERS frame <length=58, flags=0x05, stream_id=1>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; Open new stream
          :authority: api.conduit.svc.cluster.local:8086
          :method: GET
          :path: /api/v1/Report
          :scheme: http
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/0.6.4
[  0.001] recv SETTINGS frame <length=0, flags=0x00, stream_id=0>
          (niv=0)
[  0.001] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.001] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
^C
```

I can spend some time debugging what the issue is exactly. I spent a little bit of time trying to find the root cause a couple of weeks ago but wasn't able to come down to a root cause.

**Comment by briansmith** [comment]
> root@controller-799c7d4ffc-zzxmr:/# nghttp -v http://proxy-api.conduit.svc.cluster.local:8086/api/v1/Report

If I understand this correctly, you have SSH'd into a container in the controller, and you're trying to make an HTTP request `GET http://proxy-api.conduit.svc.cluster.local:8086/api/v1/Report` and that fails. OK.

Is the Conduit proxy bound to its external listening port? Did it accept the incoming connection? it seems like it must have since the TCP handshake was completed. Is the proxy itself stuck trying to find (via DNS) or connect to proxy-api.conduit.svc.cluster.local:8086? Has it connected to the underlying service (proxy-api pod, in this case)? These can be answered by `kubectl log` for the proxy container in that pod (I hope). Similarly, the `kubectl log` for the proxy-api container would be useful. Maybe proxy-api is having trouble discovering one of the internal components of the control plane?

Anyway, we're going to switch this to use `localhost` anyway, but it's still important to know why we *have* to--that is, why it is mandatory for correct operation instead of just a performance improvement.

**Comment by deebo91** [comment]
That is correct. I ssh'd into the proxy-api container and tried to issue an HTTP/2 request to the proxy endpoint. Expected behavior is that it should send a `GOAWAY` frame since I am not sending any data along with the request.
 
> Is the Conduit proxy bound to its external listening port?

This is just a guess but I am assuming this log lines
```bash
1519322727.927578 DBUG conduit_proxy::transport::so_original_dst "inbound", get_original_dst TcpStream { sys: TcpStream { inner: TcpStream { addr: V4(172.17.0.7:4143), peer: V4(172.17.0.11:50386), fd: 36 } }, selector_id: SelectorId { id: AtomicUsize(1) } }
1519322727.928783 TRCE conduit_proxy::inbound recognize local=172.17.0.7:4143 orig=Some(V4(172.17.0.7:9999))
```
indicate that the proxy is communicating with port 4143.

> Is the proxy itself stuck trying to find (via DNS) or connect to proxy-api.conduit.svc.cluster.local:8086?

The proxy seems to be communicating with DNS just fine as it is able to resolve the proxy-api's DNS name.
```bash
1519322863.034206 TRCE conduit_proxy::dns "controller-client", resolve proxy-api.conduit.svc.cluster.local
1519322863.034552 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.107.26.247
```

I plan to sync up with @seanmonstar to see if we both can figure out what the proxy is doing with the requests I am issuing. The proxy api logs don't seem to give anything actionable that we could use to debug the issue. I will keep you posted on what else I find.

**Comment by deebo91** [comment]
With help from @seanmonstar and the kubernetes docs, I was able to get this branch working on a minikube setup. While working with sean we came to the conclusion that the issue was due to some networking configuration in minikube. We tried to `nghttp` the controller proxy-api from the conduit-proxy with its `clusterIP` but that seemed to not work. When starting minikube with 

```bash
minikube start --network-plugin=kubenet
```
I was able to run the above scenario without a hitch.  [This walthrough page](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-service/) from Kubernetes helped in figuring out the solution. It mentioned something about setting up a nodes network to enable `hairpin-mode`

When looking through the minikube logs I notices the line
```bash
 Hairpin mode set to "promiscuous-bridge" but kubenet is not enabled, falling back to "hairpin-veth"
```
So I enabled the `kubenet` plugin for minikube and that seemed to do the trick.

**Comment by adleong** [comment]
Seems like this is probably: https://github.com/kubernetes/minikube/issues/1568

Given that this change breaks Conduit on minikube and minikube is one of our target platforms, I'd be more comfortable if the `localhost` change was part of this PR just so that we're not knowingly checking in broken code.

**Comment by briansmith** [comment]
OK, if I understand this correctly, in order to work in the default mode of minikube, we *have* to use "localhost", right? That is, it really isn't optional.

**Comment by briansmith** [comment]
> Seems like this is probably: kubernetes/minikube#1568

Thanks!

> Given that this change breaks Conduit on minikube and minikube is one of our target platforms, I'd be more comfortable if the localhost change was part of this PR just so that we're not knowingly checking in broken code.

I agree. Let's just merge #381 into this. (Make sure to address the recent feedback in #381.)

File: cli/cmd/inject.go [hunk 11]
```diff
@@ -220,14 +224,31 @@ func InjectYAML(in io.Reader, out io.Writer, version string) error {
...
+		*/
 		switch meta.Kind {
 		case "Deployment":
 			var deployment v1beta1.Deployment
 			err = yaml.Unmarshal(bytes, &deployment)
 			if err != nil {
 				return err
 			}
+			if deployment.Name == "controller" {
+				DNSOverride = "localhost"
```
Comment (briansmith) [inline cli/cmd/inject.go]: You need to consider deployment.Namespace too.

File: cli/cmd/install_test.go [hunk 12]
```diff
@@ -32,7 +30,8 @@ func TestRender(t *testing.T) {
 			CreatedByAnnotation:      "CreatedByAnnotation",
 		}
 
-		err = render(config, &buf)
+		buf := new(bytes.Buffer)
+		err = render(config, buf)
```
Comment (briansmith) [inline cli/cmd/install_test.go]: Undo this change (bad merge).

File: cli/cmd/install_test.go [hunk 13]
```diff
@@ -14,8 +14,6 @@ func TestRender(t *testing.T) {
 		}
 		expectedContent := string(goldenFileBytes)
 
-		var buf bytes.Buffer
-
```
Comment (briansmith) [inline cli/cmd/install_test.go]: Undo this change. (bad merge.)

File: cli/cmd/inject.go [hunk 14]
```diff
@@ -220,14 +224,31 @@ func InjectYAML(in io.Reader, out io.Writer, version string) error {
...
+		       where pods are unable to connect to themselves through their associated service IP.
+		       Setting the CONDUIT_PROXY_CONTROL_URL to localhost allows the proxy to bypass kube DNS
+		       name resolution as a workaround to this issue.
+			2. We avoid the TLS overhead in encrypting and decrypting intra-pod traffic i.e. traffic
+		       between containers in the same pod.
+			3. Using a Service IP instead of localhost would mean intra-pod traffic would be load-balanced
+		       across all controller pod replicas. This is undesirable as we would want all traffic between
+			   containers to be self contained.
+			4. We skip recording telemetry for intra-pod traffic within the control plane.
+		*/
```
Comment (briansmith) [inline cli/cmd/inject.go]: Not sure why we would use `/* ... */` here instead of `//` comments here.

File: cli/cmd/inject.go (hunk 11 above)
Comment (briansmith) [reply cli/cmd/inject.go]: Please add a test case for a deployment named "controller" that isn't in the conduit control plane namespace, to verify we don't make it "localhost' for it.

Comment (deebo91) [reply cli/cmd/inject.go]: good point. Will add a test any any others that may be necessary.

File: cli/cmd/install_test.go (hunk 12 above)
Comment (deebo91) [reply cli/cmd/install_test.go]: Good catch!

File: cli/cmd/inject.go (hunk 14 above)
Comment (deebo91) [reply cli/cmd/inject.go]: I can convert it to `//` for consistency. It seemed easier to format when using `/* ...*/`. It looks like go prefers the latter style for package level comments.

**Comment by deebo91** [comment]
@briansmith @siggy I have added two new tests that test the new functionality that distinguishes between a controller pod in the control plane namespace. I have also done some minor refactoring in `inject.go` to make it more readable.

File: cli/cmd/install_test.go [hunk 15]
```diff
@@ -14,8 +14,10 @@ func TestRender(t *testing.T) {
 		}
 		expectedContent := string(goldenFileBytes)
 
+		testNamespace := "Namespace"
+		controlPlaneNamespace = testNamespace // Set control plane namespace to enforce localhost DNS override.
```
Comment (briansmith) [inline cli/cmd/install_test.go]: I wouldn't expect this test to need to change at all.

Doesn't the setting of `controlPlaneNamespace` need to happen in install.go itself, before calling InjectYAML? In other words, isn't this wallpapering over a real bug by fixing it only during testing but not IRL?

More generally, `InjectYAML` is relying on a bunch of global variables to be set, which are set in the 'RunE` function for the `conduit inject` command, but which may not be set for the `conduit install` command. (This is what's terrible about global variables). Let's make sure that all the global variables that `InjectYAML` depends on are actually set.

File: cli/cmd/install_test.go (hunk 12 above)
Comment (briansmith) [reply cli/cmd/install_test.go]: This change still needs to be undone.

File: cli/cmd/inject_test.go [hunk 16]
```diff
@@ -18,6 +18,8 @@ func TestInjectYAML(t *testing.T) {
 		{"inject_emojivoto_deployment.input.yml", "inject_emojivoto_deployment.golden.yml"},
 		{"inject_emojivoto_deployment_hostNetwork_false.input.yml", "inject_emojivoto_deployment_hostNetwork_false.golden.yml"},
 		{"inject_emojivoto_deployment_hostNetwork_true.input.yml", "inject_emojivoto_deployment_hostNetwork_true.golden.yml"},
+		{"inject_emojivoto_deployment_controller_name.input.yml", "inject_emojivoto_deployment_controller_name.golden.yml"},
+		{"inject_conduit_deployment_insert_localhost_name.input.yml", "inject_conduit_deployment_insert_localhost_name.golden.yml"},
```
Comment (briansmith) [inline cli/cmd/inject_test.go]: NIT: I don't think the `{"inject_conduit_deployment_insert_localhost_name.input.yml", "inject_conduit_deployment_insert_localhost_name.golden.yml"}` test case is needed because we already test that case in install_test.go, so I think we could remove it.

Comment (deebo91) [reply cli/cmd/inject_test.go]: Sure I could do that. I wasn't sure if it was better to have the additional test for `inject.go` to specifically highlight this change, considering the fact that the localhost modification happens in `inject.go` rather than `install.go`.

File: cli/cmd/install_test.go (hunk 15 above)
Comment (briansmith) [reply cli/cmd/install_test.go]: Consider:

```go
func addInjectInstallSharedFlags(cmd) {
	cmd.PersistentFlags().StringVarP(&conduitVersion, "conduit-version", "v", version.Version, "tag to be used for Conduit images")
	injectCmd.PersistentFlags().StringVar(&initImage, "init-image", "gcr.io/runconduit/proxy-init", "Conduit init container image name")
	injectCmd.PersistentFlags().StringVar(&proxyImage, "proxy-image", "gcr.io/runconduit/proxy", "Conduit proxy container image name")
	injectCmd.PersistentFlags().StringVar(&imagePullPolicy, "image-pull-policy", "IfNotPresent", "Docker image pull policy")
	injectCmd.PersistentFlags().Int64Var(&proxyUID, "proxy-uid", 2102, "Run the proxy under this user ID")
	injectCmd.PersistentFlags().StringVar(&proxyLogLevel, "proxy-log-level", "warn,conduit_proxy=info", "log level for the proxy")
}
```

And then changing `init()` in `inject.go` and `install.go` to call 

`addInjectInstallSharedFlags(&injectCmd)` and  `addInjectInstallSharedFlags(&installCmd)` respectively.

That way one can use `conduit install` to configure the proxy the same way in the controller that it is being configured during `conduit inject` elsewhere, which is needed for our testing.

Comment (briansmith) [reply cli/cmd/install_test.go]: > Doesn't the setting of controlPlaneNamespace need to happen in install.go itself, before calling InjectYAML? In other words, isn't this wallpapering over a real bug by fixing it only during testing but not IRL?

No, the code is correct and I was wrong. I was misremembering how command line flags are dealt with. No change is needed here, other than needing to add something like the `addInjectInstallSharedFlags()` stuff.

Comment (deebo91) [reply cli/cmd/install_test.go]: > add something like the addInjectInstallSharedFlags() stuff.

Is it safe to assume this function will be in `root.go`? It looks like its very similar to `addControlPlaneNetworkingArgs` in `root.go`

Comment (briansmith) [reply cli/cmd/install_test.go]: > Is it safe to assume this function will be in root.go? It looks like its very similar to addControlPlaneNetworkingArgs in root.go

I would probably put it into inject.go so that everybody who adds a new flag to `conduit inject` is made away that they should consider whether it needs to be added to `conduit install` too.

File: cli/cmd/inject.go [hunk 17]
```diff
@@ -83,7 +88,7 @@ func runInjectCmd(input io.Reader, errWriter, outWriter io.Writer, version strin
  * and init-container injected. If the pod is unsuitable for having them
  * injected, return null.
  */
-func injectPodTemplateSpec(t *v1.PodTemplateSpec, version string) bool {
+func injectPodTemplateSpec(t *v1.PodTemplateSpec, controlPlaneDNSNameOverride string, version string) bool {
```
Comment (olix0r) [inline cli/cmd/inject.go]: tioli/nit: can just be `func injectPodTemplateSpec(t *v1.PodTemplateSpec, controlPlaneDNSNameOverride, version string) bool {` (no need to specify `string` twice)
